### ISSUE TITLE
minifier: treat `this` not referencing to a class instance as sideeffect free

### ISSUE THREAD

**Issue by sapphi-red** [oxc-project/oxc#21364]
To fix #21296, the minifier (and the side effect detector) now treats `this` as potentially side effect free (#21297). This is too conservative and will bail out many safe cases. By only treating `this` referencing to a class instance to have a sideeffect, those false-positives can be reduced. Whether the`this` references to a class instance should be able to determine lexically.

### cases that potentially have side effects
```js
class Foo {
  constructor() {
    this;
  }
}
```
```js
class Foo {
  constructor() {
    super();
    this;
  }
}
```
```js
class Foo {
  constructor() {
    this;
    super();
  }
}
```
```js
class Foo {
  constructor() {
    foo.call(this);
    super();
  }
}
```

### cases that are side effect free
```js
// in both module and script
this
```
```js
{
  this
}
```
```js
with(foo) {
  this
}
```
```js
class Foo {
  foo() {
    this;
  }
}
```
```js
class Foo {
  static foo() {
    this;
  }
}
```
